cpp/src/arrow/scalar.h

```diff
@@ -78,6 +78,8 @@ struct ARROW_EXPORT Scalar : public util::EqualityComparable<Scalar> {
   // TODO(bkietz) add compute::CastOptions
   Result<std::shared_ptr<Scalar>> CastTo(std::shared_ptr<DataType> to) const;
 
+  static Result<std::shared_ptr<Scalar>> FromArraySlot(const Array& array, int64_t index);
```

Comment:
I feel like this should be an Array's method more than a Scalar one, e.g. `Array::GetScalar`, complementing `GetValue` and `GetView`.

Reply:
I'll move it